torch/csrc/distributed/autograd/engine/dist_engine.cpp

```diff
@@ -73,30 +73,61 @@ class DistAccumulateGradCaptureHook
   ContextPtr autogradContext_;
 };
 
+void DistEngine::globalCpuThread(
+    const std::shared_ptr<ReadyQueue>& ready_queue) {
+  while (true) {
+    NodeTask task = ready_queue->pop();
+    if (task.isShutdownTask_) {
+      // Need to shutdown this thread.
+      C10_LOG_API_USAGE_ONCE("torch.autograd.thread_shutdown");
+      break;
+    }
+
+    auto graphTask = task.base_.lock();
+    if (graphTask == nullptr) {
+      // GraphTask has expired, ignore and continue processing.
+      continue;
+    }
+
+    // Launch the execution on a JIT thread.
+    at::launch([this,
+                graphTask,
+                graphRoot = task.fn_,
+                variables =
+                    InputBuffer::variables(std::move(task.inputs_))]() mutable {
+      InputBuffer inputs(variables.size());
+      for (size_t i = 0; i < variables.size(); i++) {
+        inputs.add(i, std::move(variables[i]), c10::nullopt, c10::nullopt);
+      }
+      execute_graph_task_until_ready_queue_empty(
+          /*node_task*/ NodeTask(graphTask, graphRoot, std::move(inputs)),
```

Comment:
Why do you have to "unpack" the given NodeTask that you poped before executing it here?

Reply:
But it was moved out of the queue right? So it could be moved to the lambda and then into the new queue?

Reply:
Yes, it can be moved to the lambda. Although, the lambda is passed to `at::launch` as `std::function<void()> func`. As a result, when the lambda is passed to `at::launch` it copies the lambda and as a result needs to copy the NodeTask (since it is part of the lambda due to the capture) and thats where it fails. If `at::launch` took the function as reference or had the function as a templated reference this wouldn't be a problem.